### PR TITLE
Polling crawls

### PR DESCRIPTION
**Pull request by ricarkol**
This is a bit of refactoring in order to make PR https://github.com/cloudviz/agentless-system-crawler/pull/52 more straightforward. The idea with this change is that there is a new `crawl` method called `polling_crawl()`. For containers (i.e. ContainersCrawler), `polling_crawl` blocks `timeout` seconds waiting for any new container, and if any, it crawls it and returns a frame. The idea with this is that if the frequency is 10 seconds, the `iterate` method of Crawler calls `polling_crawl` multiple times (once per new container if any) until 10 seconds have elapsed, after that, the regular `crawl` method is called to crawl all containers.

Guys, can you double check that the logic is OK, and that the comments make sense please?

@nadgowdas your code should be here: DockerContainer.poll_docker_containers (dockercontainer.py).

### CODE REVIEW

File: crawler/crawler.py [hunk 1]
```diff
@@ -219,14 +229,8 @@ def main():
     else:
         raise NotImplementedError('Invalid crawlmode')
 
-    emitters = EmittersManager(urls=args.url,
```
Comment (sahilsuneja1) [inline crawler/crawler.py]: I liked this one better, when crawlers and emitters were separate, it was cleaner :(

File: crawler/crawler.py [hunk 2]
```diff
@@ -195,8 +194,15 @@ def main():
...
     options['mountpoint'] = args.mountpoint
 
+    emitters = EmittersManager(urls=args.url,
+                               format=args.format,
+                               compress=args.compress,
+                               extra_metadata=args.extraMetadata)
+
     if args.crawlmode == 'OUTCONTAINER':
         crawler = ContainersCrawler(
+            emitters,
```
Comment (sahilsuneja1) [inline crawler/crawler.py]: again, emitters separate from crawlers was way better I feel!

File: crawler/containers_crawler.py [hunk 3]
```diff
@@ -68,22 +60,29 @@ def crawl_container(self, container, ignore_plugin_exception=True):
                     raise exc
         return frame
 
-    def crawl_containers(self, ignore_plugin_exception=True):
+    def polling_crawl(self, timeout, ignore_plugin_exception=True):
```
Comment (sahilsuneja1) [inline crawler/containers_crawler.py]: can this be made part of worker instead?

File: crawler/containers_crawler.py [hunk 4]
```diff
@@ -11,22 +11,25 @@ def __init__(self, feature_types, container):
         self.metadata['system_type'] = 'container'
 
 
-class ContainersCrawler(BaseCrawler):
+class ContainersCrawler(Crawler):
```
Comment (sahilsuneja1) [inline crawler/containers_crawler.py]: BaseCrawler was definitely more helpful than generic Crawler, esp. to reflect inheritance and method overwriting.

File: crawler/base_crawler.py [hunk 5]
```diff
@@ -32,20 +32,113 @@ def __str__(self):
...
-        Crawl to get a snapshot frame of all systems.
+        Crawl to get a list of snapshot frames for all systems.
 
         :param ignore_plugin_exception: ignore exceptions raised on a plugin
         :return: a list generator of Frame objects
         """
-        pass
+        raise NotImplementedError('crawl method implementation is missing.')
+
+    def polling_crawl(self, timeout, ignore_plugin_exception=True):
```
Comment (sahilsuneja1) [inline crawler/base_crawler.py]: The comment from the iterate function can be added here: "If polling is not implemented in subclass, this is a sleep(timeout)

File: crawler/crawler.py [hunk 6]
```diff
@@ -219,14 +229,8 @@ def main():
...
 
-    emitters = EmittersManager(urls=args.url,
-                               format=args.format,
-                               compress=args.compress,
-                               extra_metadata=args.extraMetadata)
-    worker = Worker(emitters, args.frequency, crawler)
-
     try:
-        worker.run()
+        crawler.run()
```
Comment (sahilsuneja1) [inline crawler/crawler.py]: worker also meant crawl + possibly filter + emit. With 'crawler' instead, that separation seems to go away.

**Comment by nadgowdas** [comment]
Overall these changes looks sufficient to implement docker container polling.

**Pull request merged**